src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
+        } else {
+            System.arraycopy(array, index, array, index + 1, size - index);
+            array[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGet(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexForSet(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForRemove(index);
+        final T removedElement = array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        array[size - 1] = null;
+        size -= 1;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < array.length; i++) {
+            if (element == null && array[i] == null
+                    || element != null && element.equals(array[i])) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException(element + "not found");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (size == 0) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private T[] resizeArray(T[] array) {
+        T[] tempArray = (T[]) new Object[(int) (array.length * RESIZE_FACTOR)];
+        System.arraycopy(array, 0, tempArray, 0, array.length);
+        array = tempArray;
+        return array;
+    }
+
+    private T[] resizeAndInsert(T[] array, int index, T value) {
+        T[] newArray = (T[]) new Object[(int) (array.length * RESIZE_FACTOR)];
+        System.arraycopy(array, 0, newArray, 0, index);
+        newArray[index] = value;
+        System.arraycopy(array, index, newArray, index + 1, size - index);
+        return newArray;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index > size");
+        }
+    }
+
+    private void checkIndexForGet(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("index > size");
+        }
+    }
+
+    private void checkIndexForSet(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index > size");
+        }
+    }
+
+    private void checkIndexForRemove(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("index > size");
+        }
```

Comment:
DRY principle violation
```suggestion
    private void checkIndex(int index) {
        if (index < 0 || index >= size) {
            throw new ArrayListIndexOutOfBoundsException("index > size");
        }
```